File: app/src/main/java/com/six/cat/sixcat/presenter/LiveContentPresenter.java
package com.six.cat.sixcat.presenter;

import com.six.cat.sixcat.RetrofitFactory;
import com.six.cat.sixcat.model.LiveBean;
import com.six.cat.sixcat.module.live.ILiveInterface;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * @author liguoying
 * @date 2018/1/12.
 */

public class LiveContentPresenter implements ILiveInterface.ILivePresenter {

    private ILiveInterface.ILiveView mView;
    private int count = 10;
    private int totalSize = 0;
    private int start = 0;

    public LiveContentPresenter(ILiveInterface.ILiveView mView) {
        this.mView = mView;
    }

    @Override
    public void loadData() {
        mView.onShowLoading();
        RetrofitFactory.getRetrofit().create(ILiveApi.class).getLiveContent("北京", start, count)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(liveBean -> {
                    totalSize = liveBean.getTotal();
                    return liveBean.getSubjects();
                })
                .compose(mView.<List<LiveBean.SubjectsBean>>bindToLife())
                .subscribe(subjectsBeans -> {
                    if (subjectsBeans.size() > 0) {
                        mView.onLoadDataSuccess(subjectsBeans, totalSize);
                    } else {
                        mView.haveNoMore();
                    }
                }, throwable -> doShowNetError());
    }

    @Override
    public void doRefresh() {
        start = 0;
        count = 10;
        mView.onShowLoading();
        loadData();
    }

    @Override
    public void doLoadMoreData() {
        start += 10;
        count = (start >= totalSize) ? totalSize % count : count;
        loadData();
    }

    @Override
    public void doShowNetError() {
        mView.onHideLoading();
        mView.onShowNetError();
    }
}

File: app/src/main/java/com/six/cat/sixcat/presenter/LiveContentPresenter.kt
package com.six.cat.sixcat.presenter

import com.hazz.kotlinmvp.net.exception.ExceptionHandle
import com.six.cat.sixcat.base.refactor.BasePresenter
import com.six.cat.sixcat.model.LiveModel
import com.six.cat.sixcat.module.live.LiveContrace
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 * @author liguoying
 * @date 2018/1/12.
 */

class LiveContentPresenter : BasePresenter<LiveContrace.View>(), LiveContrace.Presenter {

    private var count = 10
    private var totalSize = 0
    private var start = 0

    private val liveModel by lazy { LiveModel() }

    override fun loadData(city: String, start: Int, count: Int) {
        checkViewAttached()
        val disposable = liveModel.getLiveData(city, start, count)
                .map { liveBean ->
                    mRootView?.dismissLoading()
                    totalSize = liveBean.total
                    liveBean.subjects
                }
                .subscribe({ subjectsBeans ->
                    mRootView?.apply {
                        setLiveData(subjectsBeans, totalSize)
                    }
                }, { throwable ->
                    mRootView?.apply {
                        showError(ExceptionHandle.handleException(throwable), ExceptionHandle.errorCode)
                    }
                })

        addSubscription(disposable)
    }


    override fun loadMoreData() {
    }

    fun doRefresh() {
        start = 0
        count = 10
        loadData("",start,count)
    }

    fun doLoadMoreData() {
        start += 10
        count = if (start >= totalSize) totalSize % count else count
        loadData("",start,count)
    }

//    fun doShowNetError() {
//        mView!!.onHideLoading()
//        mView.onShowNetError()
//    }
}
